dsl: Introduce abstractions for multi-stage time integrators

fernanvr opened the pull request:
Implementation of the ideas of @mloubout, as discussed in the Slack channel #timestepping.

It were created two new classes:

- MultiStage: Each instance represents a combination of a PDE and its associated time integrator.

- RK: Instances encapsulate a Butcher Tableau. The class methods define specific Runge-Kutta schemes, and it includes logic to expand a single PDE into multiple stage equations according to the RK method. It only contains the classic RK44, of fourth order and four stages, as a prove of concept.

To integrate the `MultiStage` into Devito’s pipeline, we modified the `_lower(...) `function in `operator.py` by adding the line:
`expressions = cls._lower_multistage(expressions, **kwargs)`,
and `modified _sanitize_exprs(cls, expressions, **kwargs)` to recognize `MultiStage` instances.

We also created a new function:
 `_lower_multistage()`, 
in the same file. This function handles the expansion of `MultiStage` instances into their corresponding stage equations.

fernanvr committed:
implementation of multi-stage time integrators

FabioLuporini reviewed (changes requested):
Hi, many thanks for this!

As you may see from the "Files changed" tab (https://github.com/devitocodes/devito/pull/2599/files),  there are several new files that shouldn't be there. The whole .idea folder has probably been pushed erroneously. You will have to rebase your git history and erase it.

The new file `new_classes` should have a proper name and be placed in a different folder, @EdCaunt may have some ideas about that.

The test you have added has to be placed within a function `def test_....` otherwise it won't be picked up by `pytest`. And we are going to need more tests, one isn't enough.
Also, a test is a test, it shouldn't plot anything. Instead, the numerical output should be compared to either an analytical solution or to one or more precomputed ("by hand") norms.

FabioLuporini commented on devito/operator/operator.py:
```diff
@@ -271,6 +273,9 @@ def _lower(cls, expressions, **kwargs):
         kwargs['rcompile'] = cls._rcompile_wrapper(**kwargs)
 
+        # [MultiStage] -> [Eqs]
+        expressions = cls._lower_multistage(expressions, **kwargs)
```
this should be called inside `_lower_dsl`, which in turn is called within `_lower_exprs`

fernanvr replied:
I changed it to be called within `_lower_exprs`, but I couldn't find the `_lower_dsl` function...

EdCaunt commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+        for i, row in enumerate(self.a):
+            assert len(row) == i, f"Row {i} in 'a' must have {i} entries for explicit RK"
+
+    def expand_stages(self, base_eq, eq_num=0):
```
Actually, if  `MultiClass` subclassed `Eq`, then possibly this `expand_stages` method would instead be the `_evaluate` for the class?

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,79 @@
+# Set logging level for debugging
+dv.configuration['log-level'] = 'DEBUG'
+
+# Parameters
```
Wants to use `pytest` going forward. If you look at the existing tests, you will see that a lot of them are very granular, which aids bugfixing and test coverage.

It's worth noting that tests (especially the very elementary ones) can be physically nonsensical, in order to check that some aspect of the API or compiler functions as intended. For example, it would be useful to test that a range of Butcher tableaus get correctly assembled into  the corresponding timestepping schemes, even if those timestepping schemes themselves are meaningless.

Examples of simplifications used for unit tests include using 1D where possible and using trivial `Eq`s (i.e. `Eq(f, 1)`)

EdCaunt replied:
Other tests that we include for essentially every new class include ensuring that it can be pickled (see test_pickle.py) and that it can be rebuilt correctly using its `_rebuild` method (also used for pickling). This is what the `__rargs__` and `__rkwargs__` attached to various classes are for.

fernanvr replied:
I’ve been working in that direction, but I think it’s still a bit rough

EdCaunt reviewed (commented):
Lots of comments, but looks like a good first attempt at multistage timestepping!

EdCaunt commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+from sympy import Basic
+
+
+class MultiStage(Basic):
```
This would move all of the `MultiStage` API classes to `devito.types.equation`

fernanvr replied:
I think you’re right — I’ve updated the class inheritance accordingly.

EdCaunt commented on devito/operator/operator.py:
```diff
@@ -184,8 +185,9 @@ def _sanitize_exprs(cls, expressions, **kwargs):
         for i in expressions:
-            if not isinstance(i, Evaluable):
-                raise CompilationError(f"`{i!s}` is not an Evaluable object; "
+            i_check = i.eq if isinstance(i, MultiStage) else i
```
Which again, implies that `MultiStage` should subclass `Eq`

fernanvr replied:
Yes, I’ve fixed it

EdCaunt commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+
+class RK(Basic):
+    """
+    A class representing an explicit Runge-Kutta method via its Butcher tableau.
```
For example, the `MultiStage` API should be able to express a basic predictor-corrector timestepping scheme such as MacCormack as well as a Runge-Kutta scheme

ZoeLeibowitz commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,79 @@
+# Construct and run operator
+op = dv.Operator(pdes + [rec], subs=grid.spacing_map)
+op(dt=dt0, time=nt)
+
```
In the generated C code, I noticed the following:

```
        k00[x + 2][y + 2] = v[t0][x + 2][y + 2];
        k01[x + 2][y + 2] = v[t0][x + 2][y + 2];
        k02[x + 2][y + 2] = v[t0][x + 2][y + 2];
        k03[x + 2][y + 2] = v[t0][x + 2][y + 2];
```
Is this intended? It doesn't seem optimal since they are all assigned the same value

ZoeLeibowitz commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,79 @@
+src_temporal = (1 - 2 * (np.pi * f0 * (t * dt - 1/f0))**2) * sym.exp(-(np.pi * f0 * (t * dt - 1/f0))**2)
+
+# PDE system (2D acoustic)
+system_eqs = [U[1],
```
Maybe a simple heat equation would be a good example/test?

fernanvr replied:
I incorporated for the tests

mloubout reviewed (changes requested):
Thanks for the contribution @fernanvr I think it's a great start!

I'm gonna leave some generic comments to start with and will refine as it gets updated. So here is a few main comments.

1. As @EdCaunt , the `MultiStage` object should inherit from `Eq` as it represents an equation, or in this case a set of equations. It should also be moved to `devito/types/equations.py` or make a new `devito/types/multistage.py` file.

2. Instead of that generic class with a self-referencing constructor, you should instead try to have a top level abstract class then specific cases for it. I.e something like:

```python
class MultiStage(Eq):
    def __new__(cls, eq, method, target):

```

where `target` is the target field the same way as `solve` so that you can easily have forward and backward equations.

Then for RK you would have

```python
class RK(MultiStage):
    a
    b
    c
```

`a,b,c` need to as class attributes so you can easily create each case as their own class so you can differentiate them if needed. e.g. `RK4`, `RK3`, etc. This way you can also have a `MultiStage` class that is not RK but something else, like Adams-Bashforth or whatever.

```python
class RK4(RK)
    a = [0, 1/2, 1/2, 1]
    b = [1/6, 1/3, 1/3, 1/6]
    c = [0, 1/2, 1/2, 1]
```

You can also set it up as you did but the 

```
   @classmethod
    def RK44(cls):
```

needs to be moved out of the class and made into a plain function.

3. The `MultiStage` should implement `.evaluate(expand=False)` then you won't need the `lower_multistage` as the `evaluate` will directly return the list of equations needed.

4. THe "test" you have is more of an example and should be made into a tutorial. The tests you be testing each piece separately such as the plain construction of each object and the results of the evaluation and lowering of it.

5. Optionally. I would abstract it away and directly allow something like `eq= solve(pde, u.forward, method='RK4')` which would automatically create the right `MultiStage` object. That would allow for the user to easily switch between different time steppers.


Feel free to ask questions or discuss on slack as well

fernanvr committed:
Merge remote-tracking branch 'upstream/main' into multi-stage-time-integrator
the merge is necessary to maintaing the development of multistages

fernanvr committed:
Return of first PR comments

fernanvr committed:
Return of first PR comments

fernanvr committed:
Return of first PR comments

fernanvr commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+from sympy import Basic
+
+
+class MultiStage(Basic):
```
done

fernanvr commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+        self._validate()
+
+    def _validate(self):
+        assert len(self.a) == self.s, "'a' must have s rows"
```
done

fernanvr commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+        ----------
+        base_eq : Eq
+            The equation Eq(u.forward, rhs) to be expanded into RK stages.
+        eq_number : integer, optional
```
I believe it’s fixed now...

fernanvr commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+        rhs = base_eq.rhs
+        grid = u.grid
+        dt = grid.stepping_dim.spacing
+        t = grid.time_dim
```
appreciated and done

fernanvr commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+        """Classical Runge-Kutta of 4 stages and 4th order"""
+        a = [
+            [],
+            [1 / 2],
```
appreciated and done

fernanvr commented on devito/operator/operator.py:
```diff
@@ -314,6 +319,27 @@ def _specialize_exprs(cls, expressions, **kwargs):
 
+    @classmethod
+    @timed_pass(name='lowering.MultiStages')
+    def _lower_multistage(cls, expressions, **kwargs):
```
done

fernanvr commented on devito/operator/operator.py:
```diff
@@ -314,6 +319,27 @@ def _specialize_exprs(cls, expressions, **kwargs):
+
+        lowered = []
+        for i, eq in enumerate(as_tuple(expressions)):
+            if isinstance(eq, MultiStage):
```
I believe it's done now.

fernanvr commented on devito/operator/new_classes.py:
```diff
@@ -0,0 +1,109 @@
+        # Final update: u.forward = u + dt * sum(bᵢ * kᵢ)
+        u_next = u
+        for i in range(self.s):
+            u_next += self.b[i] * dt * k[i]
```
I incorporated the suggestion both in the final stage and in the intermediate stages

fernanvr commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,79 @@
+import sympy as sym
+import matplotlib.pyplot as plt
+
+import devito as dv
```
done

fernanvr commented:
Thank you all for the comprehensive review comments. They have been incredibly helpful in improving my coding skills and deepening my understanding of some of Devito’s workflows. Overall, I believe I have incorporated most of the suggestions to enhance the code. However, there are still a few points where I could use some additional guidance.

EdCaunt reviewed (commented):
More comments, but generally looking better

fernanvr committed:
2nd PR revision

fernanvr committed:
2nd PR revision

fernanvr commented on devito/ir/equations/algorithms.py:
```diff
@@ -95,6 +97,34 @@ def handle_indexed(indexed):
+        * If the object is MultiStage, it creates the stages of the method.
+    """
+    lowered = []
+    for i, eq in enumerate(as_tuple(expressions)):
```
done, I think..

fernanvr committed:
2rd PR, updating tests and suggestions of 2nd PR revision

fernanvr committed:
3rd PR, updating tests and suggestions of 2nd PR revision

EdCaunt reviewed (commented):
Some more comments. Improving steadily once again

fernanvr committed:
4th PR revision, code refining and improving tests

fernanvr committed:
4th PR revision, code refining and improving tests

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,179 @@
+        Number of stages in the RK method, inferred from `b`.
+    """
+
+    def __init__(self, a=None, b=None, c=None, **kwargs):
```
This can just be `def __init__(self, a: list[list[float | np.number]], b: list[float | np.number], c: list[float | np.number], **kwargs) -> None:`, avoiding the need for the validation function

fernanvr committed:
5th PR revision, one suggestion from EdC and improving tests

fernanvr committed:
including two more Runge-Kutta methods and improving tests: checking compatibility with pickles, and checking numerical convergence of three Runge-Kutta methods

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,357 @@
+
+        for i in range(1, degree):
+            alpha[i] = 1 / (mu * (i + 1)) * alpha[i - 1]
+            alpha[1:i] = 1 / (mu * list(range(1, i))) * alpha[:i - 1]
```
`(mu * list(range(1, i)))` does this definitely do what it is intended to do?

fernanvr replied:
changed for more simplicity, but yes, it was doing what it was intended to do.

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    # Time integration scheme
+    pdes = [solve(system_eqs_rhs[i] - U[i], U[i], method=time_int) for i in range(2)]
+
+    assert all(isinstance(i, MultiStage) for i in pdes), "Not all elements are instances of MultiStage"
```
For what it is worth, you don't necessarily need an error message on these asserts unless it isn't obvious what is being checked for

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    assert all(isinstance(i, MultiStage) for i in pdes), "Not all elements are instances of MultiStage"
+
+
+def test_multistage_op_computing_1eq(time_int='RK44'):
```
Since all of the tests in this file pertain to `MultiStage`, you can drop `multistage` from all function names within the file for concision

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    # Time integration scheme
+    pdes = [resolve_method(time_int)(U_multi_stage[i], system_eqs_rhs[i]) for i in range(2)]
+    op = Operator(pdes, subs=grid.spacing_map)
+    op(dt=0.01, time=1)
```
This should assert a norm or similar

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+
+    # Define wavefield unknowns: u (displacement) and v (velocity)
+    fun_labels = ['u_multi_stage', 'v_multi_stage']
+    U_multi_stage = [TimeFunction(name=name, grid=grid, space_order=2,
```
Use of a capital `U` here makes this look like a class, consider renaming

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+                      src_spatial * src_temporal]
+
+    # Time integration scheme
+    pdes = [resolve_method(time_int)(U_multi_stage[i], system_eqs_rhs[i]) for i in range(2)]
```
I'm still not a fan of this `resolve_method("method_name")` API. I think `MethodClass(lhs, rhs)` is far less ambiguous

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    # Time integration scheme
+    pde = [resolve_method(time_int)(u_multi_stage, eq_rhs)]
+    op = Operator(pde, subs=grid.spacing_map)
+    op(dt=0.01, time=1)
```
Again should assert a norm. Can also be consolidated with the previous test via parameterisation

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    op(dt=0.01, time=1)
+
+
+@pytest.mark.parametrize('time_int', ['RK44', 'RK32', 'RK97'])
```
Good! I would use the classes themselves rather than strings

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    pdes = [Eq(U[i].forward, solve(Eq(U[i].dt-eq_rhs[i]), U[i].forward)) for i in range(len(fun_labels))]
+    op = Operator(pdes, subs=grid.spacing_map)
+    op(dt=dt0, time=nt)
+    assert max(abs(U[0].data[0,:]-U_multi_stage[0].data[0,:]))<10**-5, "the method is not converging to the solution"
```
A better check would be to compare the L2 norms with `np.isclose`. See other tests in the codebase for examples

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+
+
+@pytest.mark.parametrize('time_int', ['RK44', 'RK32', 'RK97'])
+def test_multistage_methods_convergence(time_int):
```
Slightly misleading test name unless I'm missing something? This sounds like a convergence test, but seems to actually test whether the multistage timestepper generates a solution which matches a more simple timestepping scheme

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,276 @@
+    assert max(abs(U[0].data[0,:]-U_multi_stage[0].data[0,:]))<10**-5, "the method is not converging to the solution"
+
+
+
```
Nitpick: Too many blank lines at end of file

fernanvr committed:
changes to consider coupled Multistage equations

fernanvr committed:
Improvements of the  HORK_EXP

fernanvr committed:
Merge branch 'main' into multi-stage-time-integrator

EdCaunt reviewed (commented):
Some more comments. More granular testing is still needed

JDBetteridge commented on devito/operator/operator.py:
```diff
@@ -36,7 +36,6 @@
                           disk_layer)
 from devito.types.dimension import Thickness
 
-
```
Please run the linter (`flake8`) :slightly_smiling_face:

JDBetteridge commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,409 @@
+            f"_evaluate() must be implemented in the subclass {self.__class__.__name__}")
+
+
+class RK(MultiStage):
```
```suggestion
class RungeKutta(MultiStage):
```

fernanvr replied:
done

JDBetteridge commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,409 @@
+        return stage_eqs
+
+
+@register_method
+class RK44(RK):
```
I think `RK4` and indeed all RK methods should be _instances_ of the `RK` Class

Then you no longer need all of the boilerplate code below, which is just setting up Butcher tableau

JDBetteridge replied:
Or the coefficients should be class attributes and set by the child class

JDBetteridge commented on devito/types/multistage.py:
```diff
@@ -249,12 +264,12 @@ class RK97(RK):
 
 @register_method
-class HORK(MultiStage):
+class HORK_EXP(MultiStage):
```
`HighOrderRungeKuttaExplicit` perhaps

Be explicit!

Use `CamelCase` for class definitions and `names_with_underscores` for variables and functions.

JDBetteridge replied:
Also if we want to start distinguishing between explicit and implicit timesteppers we should be using a class hierarchy to do so

JDBetteridge commented on devito/types/multistage.py:
I think this file should be moved to somewhere like `devito/timestepping/rungekutta.py` or `devito/timestepping/explicitmultistage.py` that way additional timesteppers can be contributed as new files. (I'm thinking about implicit multistage, backward difference formulae etc...)

FabioLuporini replied:
I agree this file doesn't belong to `types/`

based on https://github.com/devitocodes/devito/pull/2599/changes#r3043562368, we might add it to `ir/dsl/rungekutta.py`

EdCaunt replied:
The objects are user-facing. Since it seems to sit atop other `devito.types` objects, I would suggest that a `devito/timestepping/multistage.py` would be in order

fernanvr replied:
I implemented EdC suggestion

JDBetteridge commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,413 @@
+    src_spatial.data[100, 100] = 1
+    import sympy as sym
+    src_temporal = sym.exp(- 100 * (t - 0.01)**2)
+    # import matplotlib.pyplot as plt
+    # import numpy as np
+    # t=np.linspace(0,2000,1000)
+    # plt.plot(t,np.exp(1 - 2 * (t - 1)**2))
```
Leftovers?

JDBetteridge reviewed (changes requested):
It would be nice (once this has been iterated a bit further and the API concertised) to see an example or demonstrative notebook that showcases the new functionality.

fernanvr commented on devito/types/multistage.py:
```diff
@@ -271,8 +286,19 @@ class HORK(MultiStage):
+    def source_derivatives(self, src_index, t, **kwargs):
+
+        # Compute the base wavelet function
+        f_deriv = [[self.src[i][1] for i in range(len(self.src))]]
```
Thanks! I went through the PR and changed a few of the range(len()) inside the loops

fernanvr commented on devito/types/multistage.py:
```diff
@@ -287,18 +313,33 @@ def ssprk_alpha(mu=1, **kwargs):
-        degree=kwargs.get('degree')
 
-        alpha = [0]*degree
+        alpha = [0]*self.deg
```
I could do that, but I thought that in cases like this is was better to use list. Isn't?

fernanvr commented on devito/types/multistage.py:
```diff
@@ -315,66 +356,52 @@ def _evaluate(self, **kwargs):
 
-        an_eq = range(len(U0))
+        # Create a temporary Array for each variable to save the time stages
+        # k = [Array(name=f'{kwargs.get('sregistry').make_name(prefix='k')}', dimensions=u[i].grid.dimensions, grid=u[i].grid, dtype=u[i].dtype) for i in range(n_eq)]
```
The idea is to be `Array`, but since there are still some problems when using `Array` I left it comment.

Related to the other comment, I think I could define the Arrays earlier and pass them as arguments. Is that what you have in mind?

fernanvr commented on devito/types/multistage.py:
```diff
@@ -315,66 +356,52 @@ def _evaluate(self, **kwargs):
         # Compute SSPRK coefficients
-        alpha = np.array(ssprk_alpha(mu, degree), dtype=np.float64)
+        mu = 1
+        alpha = self.ssprk_alpha(mu=mu)
```
you'r right, thanks!

fernanvr commented on devito/types/multistage.py:
```diff
@@ -315,66 +356,52 @@ def _evaluate(self, **kwargs):
 
         # Expansion coefficients for stability control
-        e_p = [0] * degree
+        e_p = [0] * self.deg
```
I could do that, but I thought that in cases like this is was better to use list. Isn't?

fernanvr commented on devito/types/multistage.py:
```diff
@@ -315,66 +356,52 @@ def _evaluate(self, **kwargs):
-
-        stage_eqs = []
+        stage_eqs = [Eq(k[i], u[i]) for i in range(n_eq)]
+        [stage_eqs.append(Eq(u[i].forward, u[i]*alpha[0])) for i in range(n_eq)]
```
changed

fernanvr commented on devito/types/multistage.py:
```diff
@@ -315,66 +356,52 @@ def _evaluate(self, **kwargs):
+        for i in range(1, self.deg-1):
+            [stage_eqs.append(Eq(k_old[j], k[j])) for j in range(n_eq)]
+            src_lhs, e_p = self.source_inclusion(u, k_old, src_index, src_deriv, e_p, t, dt, mu, n_eq)
+            [stage_eqs.append(Eq(k[j], k_old[j]+mu*dt*src_lhs[j])) for j in range(n_eq)]
```
done

fernanvr commented on tests/test_multistage.py:
```diff
@@ -1,34 +1,139 @@
+    # Time integration scheme
+    pdes = resolve_method(time_int)(U_multi_stage, system_eqs_rhs, source=src, degree=4)
+    op = Operator(pdes, subs=grid.spacing_map)
+    op(dt=0.001, time=2000)
```
done

fernanvr commented on tests/test_multistage.py:
```diff
@@ -1,34 +1,139 @@
+    assert np.max(np.abs(U[0].data[0,:]-U_multi_stage[0].data[0,:]))<10**-5, "the method is not converging to the solution"
+
+
+def test_multistage_coupled_op_computing_exp(time_int='HORK_EXP'):
```
I see your point; however, I think specifying the time integrator as a string is more convenient for users compared to having them import a specific class for their chosen method.

fernanvr committed:
Merge remote-tracking branch 'upstream/main' into multi-stage-time-integrator

fernanvr committed:
tuples, improved class names, extensive tests

- Restored from commit 7d526db48 (multi-stage-time-integrator-rebase)
- Clean imports and method registry system
- Improved class names: RungeKutta44, RungeKutta32, RungeKutta97, HighOrderRungeKuttaExponential
- Tuple-based coefficients implementation
- Extensive test suite improvements (568 lines)
- All flake8 fixes and code quality improvements

fernanvr committed:
improving spacing in some tests

fernanvr committed:
Add MFE time stepping Jupyter notebook

fernanvr committed:
Remove MFE_time_size.ipynb notebook

fernanvr committed:
Update multistage implementation and tests

- Enhanced HORK_EXP method with improved source term handling
- Fixed source inclusion and expansion coefficient computation
- Added test coverage for HORK methods
- Updated test structure with better parameterization

EdCaunt reviewed (commented):
Looking better with each review

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+            rhs = [rhs]
+
+        # Convert to tuples for immutability
+        lhs_tuple = tuple([i.function for i in lhs])
```
brackets removed, I let the object lhs_tuple becaouse it is used a few times

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+                         for item in source) if source is not None else None
+        obj._t = lhs_tuple[0].grid.time_dim
+        obj._dt = obj._t.spacing
+        obj._n_eq = len(lhs_tuple)
```
done

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+
+    @property
+    def eq(self):
+        """Return the full tuple of equations."""
```
done

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+
+        sregistry = kwargs.get('sregistry')
+        # Create temporary Arrays to hold each stage
+        k = [[Array(name=f'{sregistry.make_name(prefix='k')}', dimensions=self.lhs[j].grid.dimensions, grid=self.lhs[j].grid, dtype=self.lhs[j].dtype) for i in range(self.s)]
```
done

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+    b = (1/6, 1/3, 1/3, 1/6)
+    c = (0, 1/2, 1/2, 1)
+
+    def __init__(self, lhs, rhs, **kwargs):
```
done, I created a class called FixedTableauRungeKutta to encompass those methods and also de RK44

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+
+        # Build each stage
+        for i in range(1, self.deg - 1):
+            print('e_p:', e_p)
```
removed

fernanvr commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,540 @@
+            src_lhs, e_p = self.source_inclusion(self.lhs, k_old, e_p, **integration_params)
+            stage_eqs.extend([Eq(k_j, k_old_j+mu*self.dt*src_lhs_j) for k_j, k_old_j, src_lhs_j in zip(k, k_old, src_lhs)])
+            stage_eqs.extend([Eq(lhs_j.forward, lhs_j.forward+k_j*alpha[i]) for lhs_j, k_j in zip(self.lhs, k)])
+        print('e_p:', e_p)
```
done, and I also added comments for the lines

fernanvr committed:
updating small changes from EdC review on 26-03-2026

fernanvr committed:
Isolate multistage-related changes only

fernanvr committed:
merging two classes of Runge-Kutta

fernanvr committed:
Merge full multistage history while keeping clean tree

mloubout commented on devito/operations/solve.py:
```diff
@@ -15,7 +17,7 @@ class SolveError(Exception):
 
 
-def solve(eq, target, **kwargs):
+def solve(eq, target, method = None, eq_num = 0, **kwargs):
```
I don't see `eq_num` used anywhere. And you still have `method` there an check `kwargs` later

FabioLuporini replied:
not addressed yet

fernanvr committed:
fixing test_multistage file

FabioLuporini commented on devito/operator/operator.py:
```diff
@@ -340,6 +336,8 @@ def _lower_exprs(cls, expressions, **kwargs):
             * Apply substitution rules;
             * Shift indices for domain alignment.
         """
+        expressions = lower_multistage(expressions, **kwargs)
```
This should at least be called after `expand = ...`

and, perhaps, benefit from a more generic name such as `lower_timestepping`

FabioLuporini replied:
we could also move it inside a `_lower_dsl`, which internally calls `_specialize_dsl`, just like we already do for expressions/clusters/stree/iet

fernanvr replied:
That makes sense, thanks! This was part of an earlier approach that after one meeting with Devito's team was decided to be left like a plan b, so it shouldn’t actually be here. I’ll remove it from the PR to appear only the actual approach—though I agree that structuring it that way would make sense if we revisit this idea in the future and I already changed accordingly.

fernanvr replied:
Quick correction to my previous comment: I realized this part is actually still in use in the current implementation. I’ve updated it taking your suggestions into account (ordering + naming), so it should now reflect what it was intended.

fernanvr committed:
Remove devito/ir/equations/algorithms.py and devito/operator/operator.py changes from the PR

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,541 @@
+method_registry = {}
+
+
+def register_method(cls=None, *, aliases=None):
```
I still don't quite understand why you need this mapper from string to class. Why not just pass the class directly?

fernanvr replied:
The main reason is usability. Passing a string is simpler for users than importing and passing the corresponding class. It also allows to validate the input and provide a clear error message listing the available methods when an unknown name is given.

EdCaunt replied:
Passing a class has potential to be more extensible however, since users can subclass existing methods and pass them in without the need to register them

fernanvr replied:
ok, I removed the register and now the class will be used directly only.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,541 @@
+
+    def __new__(cls, lhs, rhs, degree=None, source=None, optimized_feature=None, **kwargs):
+        # Normalize to lists first lhs and rhs
+        if not isinstance(lhs, (list, tuple)):
```
We have `as_tuple` for this purpose

fernanvr replied:
You're right! and done

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,541 @@
+        return self._eq
+
+    @property
+    def lhs(self):
```
Why do you need to redefine these?

fernanvr replied:
You are right, a lot of those properties aren't needed, I cleaned several of them.

EdCaunt replied:
This still needs fixing

fernanvr replied:
The reason for those properties is to allow access to the tuple-valued left-hand sides and right-hand sides through `obj.lhs` and `obj.rhs`, instead of using `obj._lhs` and `obj._rhs`. This keeps the API similar to `Eq`, but adapted to tuples of expressions. But, if you think this doesn't matter that much, I can remove the properties and call them with `obj._lhs` and `obj._rhs`.

EdCaunt replied:
Isn't this already a property of `Eq`? I still don't quite get why you need to redefine it?

fernanvr committed:
implemented suggestions of EdC and Fabio

FabioLuporini commented on devito/operator/operator.py:
```diff
@@ -320,6 +319,22 @@ def _specialize_dsl(cls, expressions, **kwargs):
         return expressions
 
+    @classmethod
+    def _specialized_timestepping(cls, expressions, **kwargs):
```
this is unlikely to be necessary, so it can be dropped

fernanvr replied:
The classmethod was implemented answering to a [previous comment](https://github.com/devitocodes/devito/pull/2599#discussion_r3043562368) from you, to implement classmethods similar to the existing ones of expressions/clusters/trees. So... should I remove it?

FabioLuporini commented on devito/operations/solve.py:
```diff
@@ -55,9 +58,12 @@ def solve(eq, target, **kwargs):
-        return sols[0]
+        sols_temp = sols[0]
+
+    method = method if method is not None else kwargs.pop('method', None)
```
this part doesn't seem very clean to me

fernanvr replied:
ok, I changed how to pass the arguments on solve.py and the code line is simpler now.

mloubout commented on devito/ir/equations/algorithms.py:
```diff
@@ -96,6 +96,41 @@ def handle_indexed(indexed):
     return ordering
 
 
+def lower_timestepping(expressions, **kwargs):
```
Not sure why this `lower_timestepping` is needed. This just calls `_evaluate` so it will already be automatically called in `_lower_exprs` [here](https://github.com/devitocodes/devito/pull/2599/changes#diff-83455ef044ba9dcb5af18d09d8e9bc596cfa837d706177ea000d831f034c8de0R370)

fernanvr replied:
In principle, it could be changed. This behavior comes from a previous decision to pass only equation-level objects to _lower_exprs. If we skip _evaluate beforehand, _lower_exprs would receive MultiStage objects instead, which would then require an additional _evaluate call to flatten them into equations. Does that make sense?

mloubout replied:
> This behavior comes from a previous decision to pass only equation-level objects to _lower_exprs

Not sure where this was decided. `_lower_exprs` doesn't requore equation-level object but Evaluable object. For example sparse operation pass an UnevaluatedOpearation that gets evaluated. SO there is no reason to have thi machinery here, `_lower_exprs` first call is to _evaluate and as long as MultiStage implements it there is no reason for a separate pass

fernanvr replied:
Ah,ok... @mloubout , correct me if I’m wrong, but are you proposing that the call to `_lower_timestepping(expressions, **kwargs)` in the `_lower` function of `operator.py` is unnecessary? and let all the flattening to `_lower_exprs`?

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,509 @@
+    of update expressions for each stage in the integration process.
+    """
+
+    def __new__(cls, lhs, rhs, degree=None, source=None, optimized_feature=None, **kwargs):
```
Can this not use more components from `Eq.__new__`, using `super().__new__(cls, *args, **kwargs)`? It seems like at least some of this functionality should already be in there

fernanvr replied:
I'm not sure I fully understand your point. I'm already calling `super().__new__` when constructing the object. That said, I did add the `*args` part, which was indeed missing.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,509 @@
+
+    def __init__(self, lhs, rhs, a: CoeffsA = None, b: CoeffsBC = None,
+                 c: CoeffsBC = None, **kwargs) -> None:
+        self.a = a if a is not None else getattr(self, 'a', None)
```
There seem to be a lot of redundant checks for `None` here

fernanvr replied:
I see your point. The reason for those checks is intentional: the idea is to allow users to pass a, b and c at construction to pick arbitrary Runge-Kutta coefficients (so you can instantiate a custom Runge-Kutta method without creating a new subclass). That’s why the checks for if it isn't None. Do that make sense?

EdCaunt replied:
Why the `getattr`?  Surely `self.a` is not set at this point. You could just set `self._a` etc, then have an error that raises if any of them are `None`

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,357 @@
+        Returns
+        -------
+        list of Eq
+            A list of SymPy Eq objects representing:
```
Still needs to be done

fernanvr replied:
Ahhh, ok, sorry. I think something went wrong when I fixed it before. Now it's done!

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,509 @@
+            - 1 final update equation of the form `u.forward = u + dt * sum(b_i * k_i)`
+        """
+
+        sregistry = kwargs.get('sregistry')
```
You can probably grab the symbol registry from the kwargs

fernanvr replied:
Ok, I need more help in this point. I though that the line `sregistry = kwargs.get('sregistry')` already did that...

EdCaunt replied:
Must have misread this - ignore me

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,502 @@
+class Test_API:
+
+    @pytest.mark.parametrize('time_int', ['RK44', 'RK32', 'RK97'])
+    def test_pickles(self, time_int):
```
This probably wants moving to `tests/test_pickle`

fernanvr replied:
Done

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,502 @@
+
+        # Define wavefield unknowns: u (displacement) and v (velocity)
+        fun_labels = ['u', 'v']
+        u = [TimeFunction(name=name, grid=grid, space_order=2,
```
Can just be `name in ('u', 'v')`

fernanvr replied:
done and changed for all the other cases in test_multistage.py

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,502 @@
+        # Lower the multistage method - this should not raise an exception
+        lowered_eqs = lower_timestepping(pdes, sregistry=sregistry)
+
+        # Validate the lowered equations
```
More checks should probably be carried out. It may also be possible to simplify some of these tests in the cases where you are simply checking that the lowering is working correctly

fernanvr replied:
I think this subject needs a little of discussion. Because I think more complex tests could be important.

EdCaunt replied:
You always want to have the simplest test that stresses the functionality that you are trying to test. It is much better to have lots of tiny, specific tests than a few big "everything box" tests. Test complexity will ramp up as the functionality being tested gets more sophisticated.

fernanvr committed:
updates based on latest suggestions

EdCaunt commented on tests/test_pickle.py:
```diff
@@ -694,6 +696,56 @@ def test_foreign(self):
         assert new_msf.name == 'msf'
         assert new_msf.npoint == 3
 
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
```
Import these specifically at the top rather than having `mstage.` at the start of each

fernanvr replied:
done

EdCaunt commented on tests/test_pickle.py:
```diff
@@ -694,6 +696,56 @@ def test_foreign(self):
 
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
+    def test_multistage(self, time_int):
+        # Grid setup
```
Drop comment

fernanvr replied:
done

EdCaunt commented on tests/test_pickle.py:
```diff
@@ -694,6 +696,56 @@ def test_foreign(self):
+
+        extent=(10, 10)
+        shape=(3, 3)
+        grid = Grid(origin=(0, 0), extent=extent, shape=shape, dtype=np.float64)
```
Default origin, can be dropped

fernanvr replied:
done

EdCaunt commented on tests/test_pickle.py:
```diff
@@ -694,6 +696,56 @@ def test_foreign(self):
+        src_spatial.data[1, 1] = 1
+        src_temporal = (1 - 2 * (t * dt - 1)**2)
+
+        # PDE system (2D acoustic)
```
This test can be a lot more simple, perhaps even just integrating an RHS that just increments by one. In fact, there is no reason for this test to be 2D at all. This would help reduce maintenance and make the purpose of the test clearer (i.e. if you construct a multistage, can you pickle and unpickle it?). There's no need for it to be mathematically meaningful

fernanvr replied:
understood, thanks!

EdCaunt commented on tests/test_pickle.py:
```diff
@@ -694,6 +696,56 @@ def test_foreign(self):
+        print(f"Testing pickling for {time_int.__name__} method")
+        print(method.lhs, method.rhs)
+
+        with tempfile.NamedTemporaryFile(delete=False) as tmpfile:
```
Can this be refactored to avoid the need for a persistent temp file

fernanvr replied:
I change it to use an in-memory buffer instead of a persistent temp file

EdCaunt commented on tests/test_pickle.py:
```diff
@@ -694,6 +696,56 @@ def test_foreign(self):
         assert new_msf.name == 'msf'
         assert new_msf.npoint == 3
 
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
```
Parameterise over pickle libraries too (as per test below)?

fernanvr replied:
done

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+        obj.src = tuple(tuple(item)
+                         for item in source) if source is not None else None
+        obj.t = lhs_tuple[0].grid.time_dim
+        obj.dt = obj.t.spacing
```
These should be properties

fernanvr replied:
done

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+        self.b = b if b is not None else getattr(self, 'b', None)
+        self.c = c if c is not None else getattr(self, 'c', None)
+
+        if self.a is None or self.b is None or self.c is None:
```
Put this at the top of the `_init_ instead

`if any(i is None for i in (a, b, c)):`

EdCaunt replied:
However, I'm not sure they should actually be arguments - more likely this should be an abstract class with no `__init__` whatsoever

fernanvr replied:
Thank you for the suggestions. I have implemented the requested changes.

About the `lhs` property: you are right that `Eq` already defines `lhs`, but `MultiStage` represents a coupled system (multiple equations) rather than a single `Eq`. In this class, `lhs` and `rhs` are intentionally tuple-valued to expose all equation components, which are then used by the stage expansion logic. For that reason, I kept the `MultiStage` `lhs` property override. Does this makes sense?

EdCaunt commented on devito/timestepping/multistage.py:
```diff
@@ -0,0 +1,433 @@
+            raise ValueError("TableauRungeKutta requires coefficients 'a', 'b', and 'c'.")
+
+    @property
+    def s(self):
```
Docstring, given the parameter name is nondescript

fernanvr replied:
done

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+        """
+
+        sregistry = kwargs.get('sregistry')
+        # Create temporary Arrays to hold each stage
```
Do we still need to use temp arrays for the stages with Jack's PR and general compilation speedups?

fernanvr replied:
I dind't knew about that! Is this already implemented in Devito's version `4.8.24.dev2+gda56fadb5`? Because I just try this one because it seems like the latest and it still requires a lot of time the symbolic treatment.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+            k_j = []
+            for _ in range(self.s):
+                k_name = sregistry.make_name(prefix="k")
+                k_j.append(TimeFunction(name=k_name, grid=self.lhs[j].grid,
```
This is not user-facing (it is generated inside the compiler and should not be a kernel argument) so should really be an `Array`

fernanvr replied:
Yes, I agree and understand you, but the problem is that the `Array`s have some inconsistencies with the derivatives that do not appear in the `Function`s. When I tried to use `Array` before, I got errors that did not appear when I used a `Function` object. That is why I decided to use `Function` instead of `Array`.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+            for _ in range(self.s):
+                k_name = sregistry.make_name(prefix="k")
+                k_j.append(TimeFunction(name=k_name, grid=self.lhs[j].grid,
+                                        space_order=self.lhs[j].space_order, time_order=0, dtype=self.lhs[j].dtype))
```
It would make more sense for these to come from the RHS I think?

mloubout replied:
making dtype/space_order/grid properties would make this cleaner.

fernanvr replied:
I'm not sure about that. The idea is for these stages, `k_j`, to be intermediate steps in the solution for the fields on the LHS. That's why they inherit the properties of those fields.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+        u_next = [self.lhs[l] + self.dt *
+                  sum(bi * ki for bi, ki in zip(self.b, k[l])) for l in range(self.n_eq)]
+        stage_eqs.extend([Eq(self.lhs[l].forward, u_next[l])
+                         for l in range(self.n_eq)])
```
This section could use some refactoring for readability

fernanvr replied:
I did something here, and think that is easier to understand.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+    c = (0, 1/2, 1/2)
+
+class RungeKutta97(TableauRungeKutta):
+    """
```
Can these docstrings be factored somehow to reduce repetition?

fernanvr replied:
I did something in this regard

EdCaunt commented on devito/timestepping/multistage.py:
```diff
@@ -0,0 +1,433 @@
+    def source_derivatives(self, src_index, **kwargs):
+
+        # Compute the base wavelet function
+        f_deriv = [[src[1] for src in self.src]]
```
Is this a sparse source or?

fernanvr replied:
Yes, the idea is to be a spares source... Nonetheless, I cannot defined as a Devito's `source` and inject it like that, because I use spatial derivatives on it...

EdCaunt commented on devito/timestepping/multistage.py:
```diff
@@ -0,0 +1,357 @@
+            alpha[1:i] = 1 / (mu * list(range(1, i))) * alpha[:i - 1]
+            alpha[0] = 1 - sum(alpha[1:i + 1])
+
+        return alpha
```
Still not fixed

fernanvr replied:
You are right, now it is changed!

EdCaunt commented on devito/timestepping/multistage.py:
```diff
@@ -0,0 +1,433 @@
+
+        return alpha
+
+    def source_inclusion(self, current_state, stage_values, e_p, **integration_params):
```
Are these sparse sources or something else?

fernanvr replied:
They represent spare sources, but with application of spatial derivation. So I cannot (I don't now how) defined as Devito's `sources` for it to work.

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+        sregistry = kwargs.get('sregistry')
+        # Create a temporary Array for each variable to save the time stages
+        # k = [Array(name=f'{sregistry.make_name(prefix='k')}', dimensions=u[i].grid.dimensions, grid=u[i].grid, dtype=u[i].dtype) for i in range(n_eq)]
+        k = [TimeFunction(name=f'{sregistry.make_name(prefix="k")}', grid=self.lhs[i].grid,
```
Should be `Array`s. It should be possible to factor out at least some of this into a superclass

fernanvr replied:
About the `Array`s, as I explained before, they gave me error in some cases... but I still let them comment for the moment in case the error they have are solved at the end of this PR :)

Regarding the factorization of the `k` definition, I did something, but I'm not sure if that was what you had in mind...

EdCaunt commented on devito/types/multistage.py:
```diff
@@ -0,0 +1,433 @@
+        integration_params = {'mu': mu, 'src_index': src_index,
+                              'src_deriv': src_deriv, 'n_eq': self.n_eq}
+
+        # Build each stage
```
Building each stage could be a hook, which might enable a bunch of code to be factored out here.

fernanvr replied:
Ok, I did some refactoring and the code looks cleaner now

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+    dt = grid.stepping_dim.spacing
+    t = grid.time_dim
+    dx = extent[0] / (shape[0] - 1)
+    return grid, x, y, dt, t, dx
```
Slightly nonsensical ordering of parameters returned here

fernanvr replied:
ok, I chaged it a little...

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+    return tn, dt0, nt
+
+
+class Test_API:
```
Convention is no underscores - same for all classes in this file

fernanvr replied:
done

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+class Test_API:
+
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
+    def test_solve(self, time_int):
```
This test doesn't need to be mathematically meaningful and can be simplified

fernanvr replied:
done

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+class Test_lowering:
+
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
+    def test_object(self, time_int):
```
This is essentially a repeat of the previous test and not a very strong check

fernanvr replied:
I changed the previous one a little, so... this now as a little more complexity

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+        assert isinstance(
+            pdes, mstage.MultiStage), "Not all elements are instances of MultiStage"
+
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
```
If `[mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97]` gets used as a parameterisation in a bunch of places, then define it as a variable somewhere and just supply that variable to the decorator

fernanvr replied:
done

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+                      ) < 10 ** -5, "the method is not converging to the solution"
+
+    @pytest.mark.parametrize('time_int', [mstage.RungeKutta44, mstage.RungeKutta32, mstage.RungeKutta97])
+    def test_low_order_convergence(self, time_int):
```
If you do 1D wave on a string then you can test against an exact solution

fernanvr replied:
you are right! done it

EdCaunt commented on tests/test_multistage.py:
```diff
@@ -0,0 +1,453 @@
+
+class Test_HORK:
+
+    def test_coupled_op_computing_exp(self, time_int=mstage.HighOrderRungeKuttaExponential):
```
This is an overly-long class name

fernanvr replied:
changed it

EdCaunt reviewed (commented):
Again, looks better than before. Tests still need to be more granular and focussed though - try to separate API, lowering and compiler functionality, and mathematical accuracy where possible, as mixing them all together clutters the tests and makes failures harder to debug

mloubout commented on devito/operations/solve.py:
```diff
@@ -55,9 +56,12 @@ def solve(eq, target, **kwargs):
-        return sols[0]
+        sols_temp = sols[0]
+
+    method = kwargs.pop('method', None)
```
just make `method` explicit in the kwargs `solve(eq, target, method=None **kwargs):` then habe a standard if/else for the return here.

fernanvr replied:
done
